### PR TITLE
report state in device and network instance info

### PR DESCRIPTION
**Pull request by giggsoff**
Report ZDeviceState in device info and ZNetworkInstanceState in networkInstance info to controller

Signed-off-by: Petr Fedchenkov <giggsoff@gmail.com>

### CODE REVIEW

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 1]
```diff
@@ -1168,6 +1169,8 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 	// start remote attestation task
 	attestModuleStart(&zedagentCtx)
 
+	zedagentCtx.inMainLoop = true
+
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/zedagent.go:1173]: This seems early. I think we should wait at least until we have processed the initial config from the controller (basically the first inhaleConfig() being done. And potentially longer; don't know if we can easily tell that all of the app instances in the initial config have reached the Booting or Running state, while ignoring any app instances which get deployed after the initial boot.

File: pkg/pillar/cmd/zedagent/reportinfo.go [hunk 2]
```diff
@@ -825,3 +827,19 @@ func getBaseosUpdateCounter(ctx *zedagentContext) uint32 {
...
+func getState(ctx *zedagentContext) info.ZDeviceState {
+	if ctx.rebootCmd || ctx.deviceReboot {
+		return info.ZDeviceState_ZDEVICE_STATE_REBOOTING
+	}
+	if ctx.getconfigCtx != nil && ctx.getconfigCtx.updateInprogress {
+		return info.ZDeviceState_ZDEVICE_STATE_BASEOS_UPDATING
+	}
+	if ctx.maintenanceMode {
+		return info.ZDeviceState_ZDEVICE_STATE_MAINTENANCE_MODE
+	}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/reportinfo.go:840]: The order of these seems off. If maintenanceMode is set I think that takes precedence over others.

Does updateInprogress remain set until we have completed the testing/inprogress of the new image?  We should report ZDEVICE_STATE_BASEOS_UPDATING from the start of the update until the testing is done.

BASEOS_UPDATING should take precedence over ZDEVICE_STATE_REBOOTING.

Comment (giggsoff) [reply pkg/pillar/cmd/zedagent/reportinfo.go]: > The order of these seems off. If maintenanceMode is set I think that takes precedence over others.

So we will miss other states in this case, well, it is not common-used state, so, why not, done.
 
> Does updateInprogress remain set until we have completed the testing/inprogress of the new image? We should report ZDEVICE_STATE_BASEOS_UPDATING from the start of the update until the testing is done.

updateInprogress indicates inprogress, so, I add checking for updating state of other partition.

> BASEOS_UPDATING should take precedence over ZDEVICE_STATE_REBOOTING.

Done

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 1 above)
Comment (giggsoff) [reply pkg/pillar/cmd/zedagent/zedagent.go]: I can see existing comment here https://github.com/lf-edge/eve/blob/50b4b878f212e4ab974eb276e32c5ae543a3f322/api/go/info/info.pb.go#L553-L556 which indicates, that we are not check something from app instances. 
But delay it after inhaleDeviceConfig looks reasonable, done.

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/zedagent.go]: Actually @petr-zededa the current PR code sets it before calling inhale.
But let's go with this and see if we want to add some more feedback mechanisms between the different microservices later.
For instance, if we could wait until every object with a *Config also has the corresponding Status published it means that all of the initial config has propagated to all the microservices. But that's for later.

**Pull request merged**